### PR TITLE
Fix incorrect sparse add behavior when the sparse tensor has non-contiguous values

### LINKED ISSUE #17950: Incorrect sparse add when tensor has non-contiguous values
Repro:
```
import torch
indices=torch.tensor([[7, 1, 3]])
values=torch.tensor([[1., 1., 1.],
               [1., 1., 1.],
               [1., 1., 1.]])
x = torch.sparse_coo_tensor(indices, values, size=(10, 3))
values=torch.tensor(1.).expand(3, 3)
y = torch.sparse_coo_tensor(indices, values, size=(10, 3))
z = x + y

tensor(indices=tensor([[7, 1, 3]]),
       values=tensor([[2., 1., 1.],
                      [1., 1., 1.],
                      [1., 1., 1.]]),
       size=(10, 3), nnz=3, layout=torch.sparse_coo)
```

embedding_backward gives a sparse tensor with non-contiguous values so this can lead to problems when using `nn.Embedding`.

### PR DESCRIPTION
**Pull request by yf225**
Currently, this code gives incorrect result:
```python
import torch
indices=torch.tensor([[7, 1, 3]])
values=torch.tensor([[1., 1., 1.],
               [1., 1., 1.],
               [1., 1., 1.]])
x = torch.sparse_coo_tensor(indices, values, size=(10, 3))
values=torch.tensor(1.).expand(3, 3)
y = torch.sparse_coo_tensor(indices, values, size=(10, 3))
z = x + y

# Should have been all 2's in `values`
tensor(indices=tensor([[7, 1, 3]]),
       values=tensor([[2., 1., 1.],
                      [1., 1., 1.],
                      [1., 1., 1.]]),
       size=(10, 3), nnz=3, layout=torch.sparse_coo)
```

This PR fixes the bug by adding special handling for sparse tensors with non-contiguous values in the addition function (specifically, by cat'ing the indices and values together). 

This PR closes https://github.com/pytorch/pytorch/issues/17950 and https://github.com/pytorch/pytorch/issues/17919.

### CODE REVIEW

**Comment by zou3519** [comment]
```
Mar 19 16:36:10 ======================================================================
Mar 19 16:36:10 ERROR: test_sparse_ctor_getter_backward (__main__.TestAutograd)
Mar 19 16:36:10 ----------------------------------------------------------------------
Mar 19 16:36:10 Traceback (most recent call last):
Mar 19 16:36:10   File "/var/lib/jenkins/workspace/test/common_utils.py", line 120, in wrapper
Mar 19 16:36:10     fn(*args, **kwargs)
Mar 19 16:36:10   File "test_autograd.py", line 671, in test_sparse_ctor_getter_backward
Mar 19 16:36:10     test(sparse_size + dense_size, len(sparse_size), nnz, device)
Mar 19 16:36:10   File "test_autograd.py", line 653, in test
Mar 19 16:36:10     gradcheck(fn, (inp,))
Mar 19 16:36:10   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/gradcheck.py", line 247, in gradcheck
Mar 19 16:36:10     'numerical:%s\nanalytical:%s\n' % (i, j, n, a))
Mar 19 16:36:10   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/gradcheck.py", line 202, in fail_test
Mar 19 16:36:10     raise RuntimeError(msg)
Mar 19 16:36:10 RuntimeError: Jacobian mismatch for output 0 with respect to input 0,
Mar 19 16:36:10 numerical:tensor([[0.0145, 0.0000],
Mar 19 16:36:10         [0.0000, 0.0000],
Mar 19 16:36:10         [0.0145, 0.0108],
Mar 19 16:36:10         [0.0000, 0.0000],
Mar 19 16:36:10         [0.0145, 0.0108],
Mar 19 16:36:10         [0.0000, 0.0000],
Mar 19 16:36:10         [0.0145, 0.0108],
Mar 19 16:36:10         [0.0000, 0.0000],
Mar 19 16:36:10         [0.0145, 0.0108],
Mar 19 16:36:10         [0.0000, 0.0000]])
Mar 19 16:36:10 analytical:tensor([[0.0145, 0.0000],
Mar 19 16:36:10         [0.0000, 0.0108],
Mar 19 16:36:10         [0.0145, 0.0000],
Mar 19 16:36:10         [0.0000, 0.0108],
Mar 19 16:36:10         [0.0145, 0.0000],
Mar 19 16:36:10         [0.0000, 0.0108],
Mar 19 16:36:10         [0.0145, 0.0000],
Mar 19 16:36:10         [0.0000, 0.0108],
Mar 19 16:36:10         [0.0145, 0.0000],
Mar 19 16:36:10         [0.0000, 0.0108]])
```
idk if this test failure is legit

**Comment by ezyang** [comment]
It would be really helpful review if the PR message explained how exactly the problem was solved.

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp [hunk 1]
```diff
@@ -205,83 +205,30 @@ SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const S
...
 
-  get_sparse_impl(r)->set_nnz_and_narrow(r_i);
-  // TODO: I think it may be possible to track inside the loop and
-  // detect when we are uncoalesced (e.g., by observing that an
-  // index goes backwards) which may be more precise than using the
-  // coalesced flag here.  But this is easy.
-  return r._coalesced_(t_coalesced && s_coalesced);
+  LongTensor r_indices = at::cat({t_indices, s_indices}, 1);
+  Tensor r_values = at::cat({t_values, s_values}, 0);
+  alias_into_sparse(r, r_indices, r_values);
```
Comment (ezyang) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp]: If you cat'ed, don't you have to specify the output is not coalesced

**Comment by ezyang** [comment]
Can we get some benchmark numbers? I'm not sure if some of our embedding examples exercise sparse-sparse, but if it does that would be most representative.

I don't think it's necessarily wrong to switch to cat'ing the indices and values together, but I feel you could have also fixed the problem by simply switching values to use an accessor (which respects strides) rather than pointer arithmetic (which doesn't). So the algorithm change should be justified.

**Comment by gchanan** [comment]
How about only catting if the tensors aren't contiguous?  That way we only (potentially) slow down paths that were broken anyway.

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp (hunk 1 above)
Comment (gchanan) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp]: IMO we should make this a parameter on `alias_into_sparse` so people have to consider it.

Comment (yf225) [reply aten/src/ATen/native/sparse/SparseTensorMath.cpp]: `alias_into_sparse(...)` calls `set_indices_and_values_unsafe(...)` internally which always sets `coalesced_ = false`, and we expect users to call `sparse_tensor._coalesced_(...)` afterwards if they want to change the coalesce-ness of the sparse tensor. For example: https://github.com/pytorch/pytorch/blob/1c671c56c16b203ececd1ee6f70a0f6fcb9265a5/aten/src/ATen/native/sparse/SparseTensor.cpp#L457-L458

To simplify this API, we can add an `is_coalesced` parameter on `alias_into_sparse`, possibly in a separate PR.

**Comment by yf225** [comment]
@ezyang @gchanan I haven't figured out a way to make `THBlas_axpy` work with non-contiguous `values`, and I opt for cat'ing when the tensors aren't contiguous instead.

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp [hunk 2]
```diff
@@ -211,77 +211,95 @@ SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const S
...
   Tensor s_values = src._values();
-  LongTensor r_indices = at::empty({sparse_dim, max_nnz}, t_indices.options());
-  Tensor r_values = new_values_with_size_of(s_values, max_nnz).zero_();
   r.resize_as_(src);
-  get_sparse_impl(r)->set_indices_and_values_unsafe(r_indices, r_values);
 
-  int64_t blockSize = r_values.stride(0);
-  int64_t cmp, d;
-  int64_t r_i = 0, t_i = 0, s_i = 0;
+  if (s_values.is_contiguous() && t_values.is_contiguous()) {
```
Comment (yf225) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp]: There is no change in this if-branch compared to the original code - I only indented it.

File: aten/src/ATen/native/sparse/SparseTensorMath.cpp [hunk 3]
```diff
@@ -211,77 +211,95 @@ SparseTensor& add_out_sparse_cpu(SparseTensor& r, const SparseTensor& t, const S
...
-  // index goes backwards) which may be more precise than using the
-  // coalesced flag here.  But this is easy.
-  return r._coalesced_(t_coalesced && s_coalesced);
+    get_sparse_impl(r)->set_nnz_and_narrow(r_i);
+    // TODO: I think it may be possible to track inside the loop and
+    // detect when we are uncoalesced (e.g., by observing that an
+    // index goes backwards) which may be more precise than using the
+    // coalesced flag here.  But this is easy.
+    return r._coalesced_(t_coalesced && s_coalesced);
+  } else {
```
Comment (yf225) [inline aten/src/ATen/native/sparse/SparseTensorMath.cpp]: This if-branch is the actual addition.

**Review by gchanan — APPROVED** [review]

**Pull request closed without merging**